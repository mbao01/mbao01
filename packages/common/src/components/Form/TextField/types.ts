import { type ReactNode } from "react";
import { type InputProps } from "../Input/types";

export type TextFieldProps = InputProps & {
  info?: ReactNode;
  error?: string | string[] | null;
};
