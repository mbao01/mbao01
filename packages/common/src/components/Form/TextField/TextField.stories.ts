import type { Meta, StoryObj } from "@storybook/react";
import { TextField } from "./TextField";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: TextField,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
    // backgrounds: [{ name: 'dark background', value: '#000', default: true }],
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof TextField>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const TextOnlyField: Story = {
  args: {
    id: "first-name",
    placeholder: "First name",
  },
};

export const EmailField: Story = {
  args: {
    id: "email",
    type: "email",
    placeholder: "Email",
  },
};

export const PhoneField: Story = {
  args: {
    id: "phone-number",
    type: "tel",
    placeholder: "+44 789 1234 567",
  },
};

export const TextFieldWithLabel: Story = {
  args: {
    id: "first-name",
    label: "First name",
  },
};

export const TextFieldWithInfo: Story = {
  args: {
    id: "first-name",
    label: "First name",
    info: "Required",
  },
};

export const TextFieldWithError: Story = {
  args: {
    id: "first-name",
    label: "First name",
    error: "Something went wrong!",
  },
};

export const TextFieldWithErrors: Story = {
  args: {
    id: "first-name",
    label: "First name",
    error: ["First name is required", "Must be greater than 3 characters", "Something went wrong!"],
  },
};

export const DisabledTextField: Story = {
  args: {
    id: "last-name",
    disabled: true,
    "aria-label": "Last name",
  },
};
