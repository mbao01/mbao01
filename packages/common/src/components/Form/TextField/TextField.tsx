import React from "react";
import { Fieldset } from "../components/Fieldset";
import { Label } from "../components/Label";
import { Input } from "../Input";
import { type TextFieldProps } from "./types";

export const TextField = React.forwardRef<HTMLInputElement, TextFieldProps>(
  ({ id, name, info, error, label, ...props }: TextFieldProps, ref) => {
    const feedbackId = name ? `${name}-information` : undefined;

    return (
      <Fieldset>
        {Boolean(label ?? info) && (
          <Label className="flex justify-between" htmlFor={id}>
            <span>{label}</span>
            {info && <span>{info}</span>}
          </Label>
        )}
        <Input
          id={id}
          ref={ref}
          name={name}
          aria-invalid={error ? true : undefined}
          aria-describedby={feedbackId}
          {...props}
        />
        {Boolean(error) && (
          <Label className="flex flex-col items-start" htmlFor={id} id={feedbackId}>
            {(Array.isArray(error) ? error : [error]).map((e) => (
              <span className="text-red-500">{e}</span>
            ))}
          </Label>
        )}
      </Fieldset>
    );
  }
);
TextField.displayName = "TextField";
