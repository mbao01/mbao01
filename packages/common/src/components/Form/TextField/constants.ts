import { cva } from "../../../libs";

export const getTextFieldClasses = cva(
  "h-10 w-full rounded-md text-sm transition-all duration-100"
);
