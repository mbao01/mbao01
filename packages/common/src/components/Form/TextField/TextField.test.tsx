import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it } from "vitest";
import { TextField } from "./TextField";

describe("TextField", () => {
  it("renders with the correct initial state", () => {
    const { asFragment } = render(<TextField name="username" defaultValue="John Doe" />);

    expect(screen.getByDisplayValue("John Doe")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("has a label", () => {
    const { asFragment } = render(<TextField name="username" label="Username" />);

    expect(screen.getByLabelText("Username")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("has an info text", () => {
    const { asFragment } = render(
      <TextField name="username" label="Username" info="This is an info text" />
    );

    expect(screen.getByText("This is an info text")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("has an error text", () => {
    const { asFragment } = render(
      <TextField name="username" label="Username" error="This is an error text" />
    );

    expect(screen.getByText("This is an error text")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("handles user input correctly", async () => {
    const user = userEvent.setup();

    const { asFragment } = render(
      <TextField name="username" label="Username" defaultValue="John Doe" />
    );
    const textField = screen.getByDisplayValue("John Doe");

    await user.clear(textField);
    await user.type(textField, "Ayomide B.");

    expect(textField).toHaveValue("Ayomide B.");
    expect(asFragment()).toMatchSnapshot();
  });

  it("prevent user input when disabled", async () => {
    const user = userEvent.setup();

    const { asFragment } = render(
      <TextField name="username" label="Username" defaultValue="John Doe" disabled />
    );
    const textField = screen.getByLabelText("Username");

    await user.type(textField, "Ayomide B.");

    expect(textField).toHaveValue("John Doe");
    expect(asFragment()).toMatchSnapshot();
  });
});
