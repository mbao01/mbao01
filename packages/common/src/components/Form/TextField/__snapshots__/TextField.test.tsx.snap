// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextField > handles user input correctly 1`] = `
<DocumentFragment>
  <label
    class="label form-control w-full"
  >
    <div
      class="label"
    >
      <span
        class="label-text"
      >
        Username
      </span>
    </div>
    <input
      aria-describedby="username-information"
      class="input rounded-md transition-all duration-100 min-h-fit h-10"
      name="username"
      value="John Doe"
    />
  </label>
</DocumentFragment>
`;

exports[`TextField > has a label 1`] = `
<DocumentFragment>
  <label
    class="label form-control w-full"
  >
    <div
      class="label"
    >
      <span
        class="label-text"
      >
        Username
      </span>
    </div>
    <input
      aria-describedby="username-information"
      class="input rounded-md transition-all duration-100 min-h-fit h-10"
      name="username"
    />
  </label>
</DocumentFragment>
`;

exports[`TextField > has an error text 1`] = `
<DocumentFragment>
  <label
    class="label form-control w-full"
  >
    <div
      class="label"
    >
      <span
        class="label-text"
      >
        Username
      </span>
    </div>
    <input
      aria-describedby="username-information"
      aria-invalid="true"
      class="input rounded-md transition-all duration-100 min-h-fit h-10"
      name="username"
    />
    <div
      class="label"
    >
      <span
        class="label-text-alt text-red-500"
        id="username-information"
      >
        This is an error text
      </span>
    </div>
  </label>
</DocumentFragment>
`;

exports[`TextField > has an info text 1`] = `
<DocumentFragment>
  <label
    class="label form-control w-full"
  >
    <div
      class="label"
    >
      <span
        class="label-text"
      >
        Username
      </span>
      <span
        class="label-text-alt"
      >
        This is an info text
      </span>
    </div>
    <input
      aria-describedby="username-information"
      class="input rounded-md transition-all duration-100 min-h-fit h-10"
      name="username"
    />
  </label>
</DocumentFragment>
`;

exports[`TextField > prevent user input when disabled 1`] = `
<DocumentFragment>
  <label
    class="label form-control w-full"
  >
    <div
      class="label"
    >
      <span
        class="label-text"
      >
        Username
      </span>
    </div>
    <input
      aria-describedby="username-information"
      class="input rounded-md transition-all duration-100 min-h-fit h-10"
      disabled=""
      name="username"
      value="John Doe"
    />
  </label>
</DocumentFragment>
`;

exports[`TextField > renders with the correct initial state 1`] = `
<DocumentFragment>
  <label
    class="label form-control w-full"
  >
    <input
      aria-describedby="username-information"
      class="input rounded-md transition-all duration-100 min-h-fit h-10"
      name="username"
      value="John Doe"
    />
  </label>
</DocumentFragment>
`;
